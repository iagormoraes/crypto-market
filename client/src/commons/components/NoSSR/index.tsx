import React, { useState, useLayoutEffect, useEffect } from 'react';

const useEnhancedEffect =
  typeof window !== 'undefined' && process.env.NODE_ENV !== 'test'
    ? useLayoutEffect
    : useEffect;

interface Props {
  children: any;
  defer?: boolean;
  fallback?: React.ReactNode | React.ReactNode[] | null;
}

export default function NoSSR({
  children,
  defer = false,
  fallback = null,
}: Props) {
  const [isMounted, setMountedState] = useState(false);

  useEnhancedEffect(() => {
    if (!defer) setMountedState(true);
  }, [defer]);

  useEffect(() => {
    if (defer) setMountedState(true);
  }, [defer]);

  return isMounted ? children : fallback;
}
