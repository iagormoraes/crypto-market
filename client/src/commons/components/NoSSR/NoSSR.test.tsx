import { render, screen } from '@testing-library/react';

import NoSSR from '.';

describe('<NoSSR/>', () => {
  it('matches snapshot', () => {
    const { container } = render(<NoSSR>children</NoSSR>);

    expect(container).toMatchSnapshot();
  });

  it('renders only on client side', () => {
    render(
      <NoSSR>
        <p>content</p>
      </NoSSR>,
    );

    expect(screen.getByText('content')).toBeInTheDocument();
  });
});
