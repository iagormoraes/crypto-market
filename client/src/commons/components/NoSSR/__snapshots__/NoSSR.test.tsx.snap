// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoSSR/> matches snapshot 1`] = `
<div>
  children
</div>
`;
